src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt

```diff
@@ -362,8 +362,32 @@ fun processPathResolveVariants(lookup: ImplLookup, path: RsPath, isCompletion: B
             if (processor("self", base)) return true
         }
         if (processItemOrEnumVariantDeclarations(base, ns, processor, withPrivateImports = isSuperChain)) return true
-        if (base is RsTypeDeclarationElement && parent !is RsUseSpeck) {
-            // Foo::<Bar>::baz
+        if (base is RsTypeDeclarationElement && parent !is RsUseSpeck) { // Foo::<Bar>::baz
+            val shadowingProcessor = run {
+                // Self-qualified type paths `Self::Item` inside impl items are restricted to resolve
+                // to only members of the current impl or implemented trait or its parent traits
+                if (ns == TYPES && base is RsImplItem && qualifier.hasCself) {
```

Comment:
Are you sure that we want to check `ns == TYPES` instead of `TYPES in ns` here?

Reply:
Oh, it should work without it, I just forget to remove it